x-pack/plugins/ingest_manager/server/services/epm/elasticsearch/transform/install.ts

```diff
@@ -0,0 +1,173 @@
+            type: ElasticsearchAssetType.transform,
+          });
+        }
+        return acc;
+      },
+      []
+    );
+
+    // get and save transform refs before installing transforms
+    await saveInstalledEsRefs(savedObjectsClient, registryPackage.name, transformRefs);
+
+    const transforms: TransformInstallation[] = transformPathDatasets.map(
+      (transformPathDataset: TransformPathDataset) => {
+        return {
+          installationName: getTransformNameForInstallation(
+            transformPathDataset,
+            installNameSuffix
+          ),
+          content: getAsset(transformPathDataset.path).toString('utf-8'),
+        };
+      }
+    );
+
+    const installationPromises = transforms.map(async (transform) => {
+      return installTransform({ callCluster, transform });
+    });
+
+    installedTransforms = await Promise.all([Promise.all(installationPromises)]).then((results) =>
+      results.flat()
+    );
+  }
+
+  const currentInstallation = await getInstallation({
+    savedObjectsClient,
+    pkgName: registryPackage.name,
+  });
+
+  // remove the saved object reference
+  await deleteTransformRefs(
+    savedObjectsClient,
+    currentInstallation?.installed_es || [],
+    registryPackage.name,
+    previousInstalledTransformEsAssets.map((asset) => asset.id),
+    installedTransforms.map((installed) => installed.id)
+  );
+
+  return installedTransforms;
+};
+
+const isTransform = (path: string) => {
+  const pathParts = Registry.pathParts(path);
+  return pathParts.type === ElasticsearchAssetType.transform;
+};
+
+const isDatasetTransform = (path: string, datasetName: string) => {
+  const pathParts = Registry.pathParts(path);
+  return (
+    !path.endsWith('/') &&
+    pathParts.type === ElasticsearchAssetType.transform &&
+    pathParts.dataset !== undefined &&
+    datasetName === pathParts.dataset
+  );
+};
+
+async function installTransform({
+  callCluster,
+  transform,
+}: {
+  callCluster: CallESAsCurrentUser;
+  transform: TransformInstallation;
+}): Promise<EsAssetReference> {
+  await callCluster('transport.request', {
+    method: 'DELETE',
+    query: 'force=true',
+    path: `_transform/${transform.installationName}`,
+    ignore: [404],
+  });
+
+  // defer validation on put if the source index is not available
+  await callCluster('transport.request', {
+    method: 'PUT',
+    path: `_transform/${transform.installationName}`,
+    query: 'defer_validation=true',
+    body: transform.content,
+  });
+
+  await callCluster('transport.request', {
+    method: 'POST',
+    path: `_transform/${transform.installationName}/_start`,
+  });
+
+  return { id: transform.installationName, type: ElasticsearchAssetType.transform };
+}
+
+const getTransformNameForInstallation = (
+  transformDataset: TransformPathDataset,
+  suffix: string
+) => {
+  const filename = transformDataset?.path.split('/')?.pop()?.split('.')[0];
```

Comment:
can we reuse [getNameAndExtension](https://github.com/elastic/kibana/blob/master/x-pack/plugins/ingest_manager/server/services/epm/elasticsearch/ingest_pipeline/install.ts#L188) here?  Perhaps it can be moved to [here](https://github.com/elastic/kibana/blob/master/x-pack/plugins/ingest_manager/server/services/epm/elasticsearch/index.ts).

Reply:
getNameAndExtension will create a circular reference. And it is a simple piece of code not sure there value in exporting it.